tokio/src/runtime/task/harness.rs

```diff
@@ -137,16 +192,78 @@ where
         }
     }
 
+    /// Remotely abort the task.
+    ///
+    /// The caller should hold a ref-count, but we do not consume it.
+    ///
+    /// This is similar to `shutdown` except that it asks the runtime to perform
+    /// the shutdown. This is necessary to avoid the shutdown happening in the
+    /// wrong thread for non-Send tasks.
+    pub(super) fn remote_abort(self) {
+        if self.header().state.transition_to_notified_and_cancel() {
+            // The transition has created a new ref-count, which we turn into
+            // a Notified and pass to the task.
+            //
+            // Since the caller holds a ref-count, the task cannot be destroyed
+            // before the call to `schedule` returns even if the call drops the
+            // `Notified` internally.
+            self.core()
+                .scheduler
+                .schedule(Notified(self.get_new_task()));
+        }
+    }
+
     // ===== waker behavior =====
 
+    /// This call consumes a ref-count and notifies the task. This will create a
+    /// new Notified and submit it if necessary.
+    ///
+    /// The caller does not need to hold a ref-count besides the one that was
+    /// passed to this call.
     pub(super) fn wake_by_val(self) {
-        self.wake_by_ref();
-        self.drop_reference();
+        use super::state::TransitionToNotifiedByVal;
+
+        match self.header().state.transition_to_notified_by_val() {
+            TransitionToNotifiedByVal::Submit => {
+                // The caller has given us a ref-count, and the transition has
+                // created a new ref-count, so we now hold two. We turn the new
+                // ref-count Notified and pass it to the call to `schedule`.
+                //
+                // The old ref-count is retained for now to ensure that the task
+                // is not dropped during the call to `schedule` if the call
+                // drops the task it was given.
+                self.core()
+                    .scheduler
+                    .schedule(Notified(self.get_new_task()));
+
+                // Now that we have completed the call to schedule, we can
+                // release our ref-count.
+                self.drop_reference();
```

Comment:
I don't understand this bit. In theory, `wake_by_val` **should** be able to avoid the ref inc / `drop_reference` here. It *seems* like there was an explicit decision to *not* do that, but I don't understand why. The `Notified` holds a reference count so how could the task be deallocated while it is being scheduled?

Reply:
This is because `schedule` is being called on `self`, so the reference must live long enough.